File: TCPServer.py
from socket import *  # Import tất cả các thành phần từ thư viện socket

serverPort = 1200  # Cổng mà server sẽ lắng nghe kết nối

# Tạo socket TCP cho server (AF_INET: dùng IPv4, SOCK_STREAM: giao thức TCP)
serverSocket = socket(AF_INET, SOCK_STREAM)

# Gán socket với địa chỉ IP ('' nghĩa là tất cả các IP trên máy) và cổng chỉ định
serverSocket.bind(('', serverPort))

# Server bắt đầu lắng nghe kết nối, tối đa 1 kết nối đang chờ
serverSocket.listen(1)

print('The server is ready to receive')  # In ra thông báo khi server sẵn sàng

# Vòng lặp vô hạn để server luôn sẵn sàng xử lý client
while True:
    # Chấp nhận kết nối từ client, trả về socket kết nối và địa chỉ client
    connectionSocket, addr = serverSocket.accept()

    # Nhận dữ liệu từ client (tối đa 1024 byte), sau đó giải mã từ byte về chuỗi
    sentence = connectionSocket.recv(1024).decode()

    # Chuyển toàn bộ chuỗi thành chữ in hoa
    capitalizedSentence = sentence.upper()

    # Gửi chuỗi đã xử lý (chữ in hoa) lại cho client (mã hoá thành byte)
    connectionSocket.send(capitalizedSentence.encode())

    # Đóng kết nối với client hiện tại
    connectionSocket.close()

File: TCPServer.java
import java.io.*;
import java.net.*;

class TCPServer {

    public static void main(String argv[]) throws Exception {
        String clientSentence;         // Biến lưu chuỗi nhận từ client
        String capitalizedSentence;   // Biến lưu chuỗi đã chuyển thành chữ in hoa để gửi lại client

        // Tạo socket server lắng nghe tại cổng 6789
        ServerSocket welcomeSocket = new ServerSocket(6789);

        System.out.println("Server is running and waiting for connection...");

        while (true) {
            // Chấp nhận kết nối từ client (chặn cho đến khi có client kết nối)
            Socket connectionSocket = welcomeSocket.accept();

            // Tạo luồng để đọc dữ liệu từ client
            BufferedReader inFromClient = new BufferedReader(new InputStreamReader(connectionSocket.getInputStream()));

            // Tạo luồng để gửi dữ liệu về cho client
            DataOutputStream outToClient = new DataOutputStream(connectionSocket.getOutputStream());

            // Đọc chuỗi gửi từ client
            clientSentence = inFromClient.readLine();
            System.out.println("Received: " + clientSentence); // In chuỗi đã nhận ra màn hình

            // Chuyển chuỗi thành chữ in hoa và thêm dấu xuống dòng
            capitalizedSentence = clientSentence.toUpperCase() + '\n';

            // Gửi chuỗi đã xử lý lại cho client
            outToClient.writeBytes(capitalizedSentence);
        }
    }
}
